libs/git/git.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2024 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+
+	go_git "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/gravitational/trace"
+)
+
+// Repo provides a collection of functions to query and modify a single git repository.
+// Wrapper around the go-git library that also includes methods to execute git commands
+// on the system for missing compatability.
+type Repo struct {
+	dir        string
+	Repository *go_git.Repository
+}
+
+// NewRepoFromDirectory initializes [Repo] from a directory.
+func NewRepoFromDirectory(dir string) (*Repo, error) {
+	inner, err := go_git.PlainOpen(dir)
+	if err != nil {
+		return &Repo{}, trace.Wrap(err)
+	}
+
+	return &Repo{
+		dir:        dir,
+		Repository: inner,
+	}, nil
+}
+
+// GetCommitForTag attempts to get the resolved commit for a given tag.
+func (r *Repo) GetCommitForTag(tag string) (*object.Commit, error) {
+	t, err := r.Repository.Tag(tag)
+	if err != nil {
+		return &object.Commit{}, trace.Wrap(err, fmt.Sprintf("tag %q not found, use a valid git tag", tag))
+	}
+	return r.Repository.CommitObject(t.Hash())
+}
+
+// GetCommitForHead attempts to get the resolved commit for head.
+func (r *Repo) GetCommitForHead() (*object.Commit, error) {
+	ref, err := r.Repository.Reference(plumbing.HEAD, false)
+	if err != nil {
+		return &object.Commit{}, trace.Wrap(err, "can't get latest reference for ent")
+	}
+	return r.Repository.CommitObject(ref.Hash())
+}
+
+// GetBranchForHead attempts to get the resolved branch of HEAD.
+func (r *Repo) GetBranchNameForHead() (string, error) {
+	// symbolic-ref HEAD
+	ref, err := r.Repository.Reference(plumbing.HEAD, true)
+	if err != nil {
+		return "", trace.Wrap(err, "not on a branch")
+	}
+
+	if !ref.Name().IsBranch() {
+		return "", trace.BadParameter("not on a branch: %s", ref)
+	}
+	return ref.Name().String(), nil
+}
+
+// GetParentReleaseBranch will attempt to find a parent branch for HEAD.
+// This will also work if HEAD has branched from an earlier commit in a release branch e.g.
+//
+//	 o---o---HEAD
+//	/---o---o---branch/v16
+func (r *Repo) GetParentReleaseBranch() (string, error) {
+	forkPointRef, err := r.RunCmd("merge-base", "--fork-point", "HEAD")
```

Comment:
It's kind of sad that we now have to use both git library and git binary. It feels like it's the worst of both worlds tbh because we still need to have git binary in path, but now also have an extra dependency.

Sorry to keep going back and forth on this, but should we revert back to just calling git CLI?

Reply:
No worries I did kind of have some misgivings on this. I was intending this to be as close to a 1-to-1 parity of the original `changelog.sh` as possible but it does end up with code that doesn't feel so clean. Unfortunately this part is not supported out of the box for the `go-git` library and might be possible to implement but would be complex.

I was thinking a better approach could be to have this be less opinionated on how this is obtained and just expect more from the arguments. In this case the git CLI just does a better job so we could just simplify the `changelog.sh` to do some of the trickier git stuff and pass it in as an argument to this tool.

I also did think maybe git is just too good at being git so maybe we could keep git commands in the bash script. This *does* much run faster though when it doesn't have to execute the git CLI.

Reply:
Actually just thought of another solution. GitHub API does provide some methods for [querying git objects](https://docs.github.com/en/rest/git?apiVersion=2022-11-28). I think it should be enough for this use case. With that this changelog tool wouldn't need a dependency on `git` (and therefore the filesystem) at all. Any git features we do need can be passed in as arguments.

This would change the API of this changelog tool
```
changelog <base_branch> <base_tag>
```
Where `base_branch` and `base_tag` are now required arguments. Expected usage would still be `make changelog`.